### ISSUE TITLE
lightning might stuck for hours when importing parquet files from cloud storage

### ISSUE THREAD

**Issue by D3Hunter** [pingcap/tidb#56104]
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

after https://github.com/pingcap/tidb/pull/46984, including `>= v7.5.x`, `7.1.3+`, we will sample parquet all the time in serial, it's very slow and might takes hours if user have large mount of parquet files before lightning start doing import, and the time takes to sample the files might even longer than real import work.

we only need this size for displaying progress more accurately and use it as a reference when splitting engine, but slowing import this much is un-acceptable.

### 2. What did you expect to see? (Required)
start import fast
### 3. What did you see instead (Required)
it might takes hours before start doing any work

**Comment by D3Hunter** [comment]
it's a enhance type of bug, it need to pick back

**Comment by zeminzhou** [comment]
How about sampling just one file? Sample the first file and calculate a compression ratio, and use this compression ratio to estimate the remaining files.

**Comment by D3Hunter** [comment]
lgtm

**Comment by lance6716** [comment]
> How about sampling just one file

@zeminzhou Do you have real use case to support this idea? I'm afraid that different table structure may have different compression ratio, so IMO we should at least sample once for every table. But the performance is not improved if every table only has 1 parquet file.

**Comment by zeminzhou** [comment]
Good catch. The different table structure may have different compression ratio, I will fix #56205 and test with different tables.

**Comment by kennytm** [comment]
I am reopening this issue. The PR #56205 to fix `constructFileInfo` is entirely wrong and does not improve the situation much.

The issue is that ***OPENING A PARQUET FILE IN S3 IS EXTREMELY TIME-COSTLY*** (i.e. `ReadParquetFileRowCountByFile` must be avoided like plague). The PR #56205 changes the original sampling algorithm:

* for each parquet file:
    1. open it ( :-1: )
    2. read the real size
    3. set it as the file's RealSize

to this:

* for each parquet file:
    1. if the table's `sampledParquetRowSizes` is unknown (0),
        1. open it ( :ok: )
        2. read the (real size / count)
        3. store it into the table's `sampledParquetRowSizes`
   2. now that the table's `sampledParquetRowSizes` is known,
        1. open the file again (👎👎👎👎👎👎)
        2. read the count
        3. set (`sampledParquetRowSizes` * count) as the file's RealSize

https://github.com/pingcap/tidb/blob/fc8bdb54c6061fdb0af15c28943d41e5f12d9f0d/pkg/lightning/mydump/loader.go#L540-L566

So you are still reading N files in total before importing. Nothing has changed!

Instead of sampling the ratio as (real size / count), you should use the ratio (real size / file size), because getting the file size is free. Then the other file's RealSize can be calculated just by ((sampled real size / file size) * current file size).

---

I also demand we add a failpoint-enabled integration test, which explodes after opening/downloading files 49 times from `data-source-dir` (i.e. total count of `ReadFile` + `Open`). The test should consist of
* 1 database (`test-schema-create.sql`)
* 4 tables (`test.{a,b,c,d}-schema.sql`)
    * two table with 10 parquet files (`test.{a,b}.[0-9].parquet`)
    * two table with 10 gzip-compressed CSV files (`test.{c,d}.[0-9].csv.gz`)

**Comment by kennytm** [comment]
@D3Hunter 

> it still need to open parquet file for row-count like original implementation, see [#56205 (review)](https://github.com/pingcap/tidb/pull/56205#pullrequestreview-2441340662) too

Yes and this is unacceptable, our customer on v7.5.5 is still waiting 3 hours to get to the pre-check stage because there are 10000 parquet files. 

Lightning should never need to open the data files --- at least not in O(N) scale --- before actually running import.

**Comment by D3Hunter** [comment]
> [@D3Hunter](https://github.com/D3Hunter)
> 
> > it still need to open parquet file for row-count like original implementation, see [#56205 (review)](https://github.com/pingcap/tidb/pull/56205#pullrequestreview-2441340662) too
> 
> Yes and this is unacceptable, our customer on v7.5.5 is still waiting 3 hours to get to the pre-check stage because there are 10000 parquet files.
> 
> Lightning should never need to open the data files --- at least not in O(N) scale --- before actually running import.

lightning need row-count to generate row-id range, it's a must step. unlike CSV/SQL, parquet might have larger file size difference for same row count due to its column format nature, might be harder to get the id-range by sample.

but we can do it in parallel to speedup

**Comment by kennytm** [comment]
@D3Hunter You only need an ***upper estimate*** of the row count, you don't need the exact count. Row ID gaps are expected. You could even set `rowCount = realSize`. 

We won't accept solutions that open and read 10000 parquet files before precheck, even with parallelization.

**Comment by D3Hunter** [comment]
> [@D3Hunter](https://github.com/D3Hunter) You only need an _**upper estimate**_ of the row count, you don't need the exact count. Row ID gaps are expected. You could even set `rowCount = realSize`.

as `realSize` comes from sampling some files, not sure how much it can reflect all files, but it should be fine in most cases

the auto-id range is not always needed, such as when the pk is clustered and data files have provided the auto IDs, maybe we can add auto detect and avoid alloc id range.

> We won't accept solutions that open and read 10000 parquet files before precheck, even with parallelization.

even with sampling, in worst case when table-count ~ O(n) where n is number of files, we still need to read O(n) files

**Comment by kennytm** [comment]
> the auto-id range is not always needed, such as when the pk is clustered and data files have provided the auto IDs, maybe we can add auto detect and avoid alloc id range.

yeah allocating ID at real-time (like TiDB) instead of pre-allocating can also get rid of the estimation, though this would be a quite big refactoring 🙃, and also not so friendly for checkpoints.

> even with sampling, in worst case when table-count ~ O(n) where n is number of files, we still need to read O(n) files

that's true, ideally we should never sampling anything at all. but in practice the number of tables is much less than the number of data files. having (number of files / number of tables >> number of CPU cores) is enough to outperform parallelization.

**Comment by joechenrh** [comment]
If opening files with parallelization is still unacceptable, can we estimate a **lower bound** for the parquet compression ratio?

I did a simple test: I created a extremely simple table, only one column and all values were 1. I used Python PyArrow package to write it as Parquet file. The data page size was set to 1MB and the number of row groups was 1 ([data page](https://tajo.apache.org/docs/0.8.0/table_management/parquet.html) is the basic unit of compression which affects the compression ratio).  Below is the result:

![Image](https://github.com/user-attachments/assets/5f9b0a2a-db9d-4d36-a335-9d4f1d0ab09b)

By this way, we can obtain the compression ratio: 20000 / (800m * 8) = 3.125e-06. Then we could use this value to estimate the `Rows` of each file, although we will quite a very large RowID gap 🧐.

> I tested different compression algorithms supported by Parquet. The file size is almost the same so I just take the minimum value.

**Comment by joechenrh** [comment]
Or we just skip fetching row count and use the above estimation method only when table has no auto row ID 🥲.
https://github.com/pingcap/tidb/blob/4734c9c2f4a240818bb2d7aac84a4c169318b6e9/pkg/lightning/backend/kv/sql2kv.go#L216-L241

**Comment by kennytm** [comment]
Using 1/320000 as the lower-bound is fine, but that puts an upper limit of the total supported size of Parquet files to (2<sup>64</sup> / 320000) = 52 TiB.

In the long term we should have some way to recover from using rowID outside of the allocated range (the issue is already happening with CSV, see TICKET-5411) and then we can reduce the compression estimation to some more practical value (e.g. 0.7). But that involves refactoring outside the scope of this issue.

BTW does `IMPORT INTO` have the same issue regarding row ID sampling?

**Comment by joechenrh** [comment]
> BTW does IMPORT INTO have the same issue regarding row ID sampling?

`IMPORT INTO` will read row number from file because we doesn't set `Rows` in  file meta.

https://github.com/pingcap/tidb/blob/4734c9c2f4a240818bb2d7aac84a4c169318b6e9/pkg/executor/importer/table_import.go#L402

https://github.com/pingcap/tidb/blob/4734c9c2f4a240818bb2d7aac84a4c169318b6e9/pkg/lightning/mydump/region.go#L364-L377

**Comment by D3Hunter** [comment]
> * rows [15, 30) of chunk 1 get ID [510, 525) <- 500 is the max rowID of the entire table

it's more complex when using parallel import where we use a table to coordinate id allocation, and also considering the range is also saved in checkpoint

reading files in parallel is simpler, if it can finish within a reasonable time, I think it's acceptable for the short term. we can consider optimize id allocation later, including the similar case for CSV
